src/shogun/mathematics/linalg/LinalgNamespace.h

```diff
@@ -0,0 +1,112 @@
+ *     this list of conditions and the following disclaimer.
+ *
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *
+ *  3. Neither the name of the copyright holder nor the names of its
+ *     contributors may be used to endorse or promote products derived from
+ *     this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Authors: 2016 Pan Deng, Soumyajit De, Heiko Strathmann, Viktor Gal
+ */
+
+#include <shogun/mathematics/linalg/LinalgBackendBase.h>
+#include <shogun/mathematics/linalg/SGLinalg.h>
+
+namespace shogun
+{
+
+namespace linalg
+{
+
+/**
+ * Wrapper method for internal implementation of vector dot-product that works
+ * with generic vectors.
+ *
+ * @param a first vector
+ * @param b second vector
+ * @return the dot product of \f$\mathbf{a}\f$ and \f$\mathbf{b}\f$, represented
+ * as \f$\sum_i a_i b_i\f$
+ */
+template <typename Type>
+Type dot(const SGVector<Type>& a, const SGVector<Type>& b)
+{
+	if (a.on_gpu() && b.on_gpu())
+	{
+		return sg_linalg->get_gpu_backend()->dot(a, b);
+	}
+	else if (a.on_gpu() || b.on_gpu())
+		SG_SERROR("Cannot operate with one vector on CPU while the other one is on GPU.\n");
+	return sg_linalg->get_cpu_backend()->dot(a, b);
+}
+
+/**
+ * Wrapper method for transferring vector from CPU to GPU
+ *
+ * @param vector SGVector to be transferred
+ * @return SGVector with vector on GPU if GPU backend is available
+ * and a shallow-copy of SGVector with vector on CPU if GPU backend not available
+ */
+template <typename T>
+SGVector<T> to_gpu(const SGVector<T>& vector)
+{
+	REQUIRE(!vector.on_gpu(), "The vector is already on GPU.\n");
+	LinalgBackendBase* gpu_backend = sg_linalg->get_gpu_backend();
+	if (gpu_backend)
+		return SGVector<T>(gpu_backend->to_gpu(vector), vector.vlen);
+	else
+	{
+		SG_SWARNING("Trying to run GPU code without GPU backend registered.\n");
+		return vector;
+	}
+}
+
+/**
+ * Wrapper method for transferring vector from GPU to CPU
```

Comment:
Dont talk about wrappers, say what it does, see above.
Also dont say "to CPU" as above, but rather: "Fetches data from GPU memory"